Examples/MAX78000/Flash_CLI/main.c

```diff
@@ -200,26 +260,42 @@ void vCmdLineTask(void *pvParameters)
                 } else if ((tmp == '\r') || (tmp == '\n')) {
                     printf("\r\n");
                     /* Null terminate for safety */
-                    buffer[index] = 0x00;
+                    buffer[index] = NULL_TERMINATION;
+                    checkLeadingSpaces(buffer, &index);
                     /* Evaluate */
-                    do {
-                        xMore = FreeRTOS_CLIProcessCommand(buffer, output, OUTPUT_BUF_SIZE);
-                        /* If xMore == pdTRUE, then output buffer contains no null
-						 * termination, so we know it is OUTPUT_BUF_SIZE. If pdFALSE, we can
-						 * use strlen.
-						 */
-                        for (x = 0; x < (xMore == pdTRUE ? OUTPUT_BUF_SIZE : strlen(output)); x++) {
-                            putchar(*(output + x));
-                        }
-                    } while (xMore != pdFALSE);
+                    if (index != 0) {
+                        do {
+                            xMore = FreeRTOS_CLIProcessCommand(buffer, output, OUTPUT_BUF_SIZE);
+                            /* If xMore == pdTRUE, then output buffer contains no null
+						    * termination, so we know it is OUTPUT_BUF_SIZE. If pdFALSE, we can
+						    * use strlen.
+						    */
+                            for (x = 0; x < (xMore == pdTRUE ? OUTPUT_BUF_SIZE : strlen(output));
+                                 x++) {
+                                putchar(*(output + x));
+                            }
+                        } while (xMore != pdFALSE);
+                    }
                     /* New prompt */
                     index = 0;
                     printf("\ncmd> ");
                     fflush(stdout);
                 } else if (index < CMD_LINE_BUF_SIZE) {
-                    putchar(tmp);
                     buffer[index++] = tmp;
-                    fflush(stdout);
+                    bool arrowKey = false;
+                    if (index >= 3) {
+                        arrowKey = checkArrowKeys(buffer + index - 3);
+                    }
+                    if (!arrowKey) {
+                        /* Echo out if not an arrow key */
+                        if ((tmp != ARROW_KEY_CODE_1) && (tmp != ARROW_KEY_CODE_2))
+                            putchar(tmp);
```

Comment:
I think we can avoid the second check on line 291 if we reorder things slightly.

```C
bool arrowKey = false;
if (index >= 3) {
    arrowKey = checkArrowKeys(buffer + index - 3);
    if (arrowKey) { //Remove from buffer }
}
if (!arrowKey) {
    putchar(tmp);
    fflush(stdout);
}

Reply:
That check is needed because over UART we receive only one byte of data through interrupts. For arrow inputs, we receive three bytes of data, which eventually triggers the interrupt three times. This would lead to few more undesirable echo statements.

And so, I included it, as a part of sanity check.  

NOTE:- This example code does not support **[** input, so custom commands shall not use this symbol.